Src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs

```diff
@@ -337,5 +338,137 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
+        {
+            if (task == null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (!cancellationToken.CanBeCanceled || task.IsCompleted)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return SingletonTask<int>.CanceledTask;
+            }
+
+            return WithCancellationSlow(task, cancellationToken);
+        }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        private static async Task<T> WithCancellationSlow<T>(Task<T> task, CancellationToken cancellationToken)
+        {
+            Debug.Assert(task != null);
+            Debug.Assert(cancellationToken.CanBeCanceled);
+
+            var tcs = new TaskCompletionSource<bool>();
+            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
+            {
+                if (task != await Task.WhenAny(task, tcs.Task).ConfigureAwait(false))
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                }
+            }
+
+            // Rethrow any fault/cancellation exception, even if we awaited above.
+            // But if we skipped the above if branch, this will actually yield
+            // on an incompleted task.
+            return await task.ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        private static async Task WithCancellationSlow(this Task task, CancellationToken cancellationToken)
+        {
+            Debug.Assert(task != null);
+            Debug.Assert(cancellationToken.CanBeCanceled);
+
+            var tcs = new TaskCompletionSource<bool>();
+            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
+            {
+                if (task != await Task.WhenAny(task, tcs.Task).ConfigureAwait(false))
```

Comment:
The reason is that this method should not throw `OperationCanceledException` if `task` has already completed. The `task` may have completed successfully or faulted, or even already been canceled itself. In any of those cases, we should return the result of the completed task rather than throwing as if the operation were canceled before it finished.
I believe TPL and other .NET code follow the same guidelines, and I am trying to be consistent with that: don't report an operation is canceled if it has already completed.